scrapy/core/downloader/handlers/aiohttp.py

```diff
@@ -0,0 +1,52 @@
+"""Download handlers for http and https schemes"""
+import logging
+import asyncio
+
+import aiohttp
+from twisted.internet import defer
+
+from scrapy.http import Headers
+from scrapy.responsetypes import responsetypes
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPDownloadHandler(object):
+
+    def __init__(self, settings):
+        self.settings = settings
+
+    def download_request(self, request, spider):
+        return _force_deferred(self._download_request(request, spider))
+
+    def _download_request(self, request, spider):
+        """Return a deferred for the HTTP download"""
+        headers=list((k.decode('latin1'), v.decode('latin1'))
+                     for k, vs in request.headers.items()
+                     for v in vs)
+
+        aioresponse = yield from aiohttp.request(
```

Comment:
If you work with asyncio in native way you almost never use
`asyncio.Future` but coroutines and `yield from` expression.

Callback chain from `Deferred` looks powerful but too error-prone. I cannot
count how many times I had errors because I forgot returning value for next
processor in the chain or finishing chain with errback guard.

On Mon, Aug 31, 2015 at 1:29 PM Daniel Graña <notifications@github.com>
wrote:

> In scrapy/core/downloader/handlers/aiohttp.py
> <https://github.com/scrapy/scrapy/pull/1455#discussion_r38335703>:
>
> > +
> > +class HTTPDownloadHandler(object):
> > +
> > +    def __init__(self, settings):
> > +        self.settings = settings
> > +
> > +    def download_request(self, request, spider):
> > +        return _force_deferred(self._download_request(request, spider))
> > +
> > +    def _download_request(self, request, spider):
> > +        """Return a deferred for the HTTP download"""
> > +        headers=list((k.decode('latin1'), v.decode('latin1'))
> > +                     for k, vs in request.headers.items()
> > +                     for v in vs)
> > +
> > +        aioresponse = yield from aiohttp.request(
>
> that's right, Deferred can be extracted from Twisted and I find them much
> more feature rich than asyncio Futures, although if we go all-in for
> asyncio the preferred idiom would be coroutines instead of callbacks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scrapy/scrapy/pull/1455/files#r38335703>.
>
-- 
Thanks,
Andrew Svetlov